tt_eager/tt_dnn/op_library/moreh_norm/moreh_norm_op.cpp

```diff
@@ -28,29 +28,84 @@ namespace primary {
-    } else if (dim == input_rank - 2) {
-        output_shape[input_rank - 2] = TILE_HEIGHT;
-        output_padding[input_rank - 2] = Padding::PadDimension{0, TILE_HEIGHT - 1};
-    } else {
-        output_shape[dim] = 1;
-    }
 
-    return Shape(output_shape, output_padding);
-}
+inline void validate_output_tensor_with_keepdim(const Tensor& input, const Tensor& output, const int64_t &dim, const bool &keepdim) {
+    auto input_shape = input.get_legacy_shape();
+    auto input_shape_wo_padding = input_shape.without_padding();
+    const auto input_rank = input_shape.rank();
+
+    const auto& output_shape = output.get_legacy_shape();
+    const auto& output_shape_wo_padding = output_shape.without_padding();
+    const auto output_rank = output_shape.rank();
+
+    const bool is_tile_dim = (dim == input_rank - 1 || dim == input_rank - 2);
+
+    log_debug(LogOp, "{}:{} keepdim {} dim {}", __func__, __LINE__, keepdim, dim);
+    log_debug(LogOp, "{}:{} input_shape {} wo_padding {}", __func__, __LINE__, input_shape, input_shape_wo_padding);
+    log_debug(LogOp, "{}:{} output_shape {} wo_paddoutg {}", __func__, __LINE__, output_shape, output_shape_wo_padding);
+
+    if (keepdim) {
+        bool ranks_are_equal = (input_rank == output_rank);
+        input_shape[dim] = (is_tile_dim) ? (TILE_HEIGHT) : (1);
+        input_shape_wo_padding[dim] = 1;
 
-inline Tensor create_output_tensor(const Tensor &input, int64_t dim, const MemoryConfig &output_mem_config) {
-    const auto output_shape = compute_output_shape(input.get_legacy_shape(), dim);
-    const auto &output =
-        create_device_tensor(output_shape, input.get_dtype(), Layout::TILE, input.device(), output_mem_config);
-    return std::move(output);
+        if (!ranks_are_equal) {
+            log_warning(
+                LogOp,
+                "{}:{} input_rank {} and output_rank {} are not the same in keepdim mode",
+                __func__,
+                __LINE__,
+                input_rank,
+                output_rank);
+        }
+
+        std::vector<uint32_t> input_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+        std::vector<uint32_t> output_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+        std::vector<uint32_t> input_dim_wo_padding(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+        std::vector<uint32_t> output_dim_wo_padding(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+        expand_to_max_dim(input_dim, input_shape);
+        expand_to_max_dim(output_dim, output_shape);
+        expand_to_max_dim(input_dim_wo_padding, input_shape_wo_padding);
+        expand_to_max_dim(output_dim_wo_padding, output_shape_wo_padding);
+
+        for (int i = 0; i < input_rank; ++i) {
+            TT_FATAL(input_dim[i] == output_dim[i]);
+            TT_FATAL(input_dim_wo_padding[i] == output_dim_wo_padding[i]);
+        }
+    } else {
+        std::vector<uint32_t> expected_output_shape;
+        std::vector<uint32_t> expected_output_shape_wo_padding;
```

Comment:
It would be nice to have such a check.

```suggestion
    } else {
        if (is_tile_dim) {
          throw error;
        }
        std::vector<uint32_t> expected_output_shape;
        std::vector<uint32_t> expected_output_shape_wo_padding;
```

Reply:
I added `TT_FATAL(!is_tile_dim);.`